torchmetrics/functional/pairwise/cosine.py

```diff
@@ -0,0 +1,85 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Optional
+
+import torch
+from torch import Tensor
+
+from torchmetrics.functional.pairwise.euclidean import _check_input, _pairwise_euclidean_distance_compute
+
+
+def _pairwise_cosine_similarity_update(
+    X: Tensor, Y: Optional[Tensor] = None, zero_diagonal: Optional[bool] = None
+) -> Tensor:
+    """Calculates the pairwise cosine similarity matrix.
+
+    Args:
+        X: tensor of shape ``[N,d]``
+        Y: if provided, a tensor of shape ``[M,d]``
+        zero_diagonal: determines if the diagonal should be set to zero
+    """
+    X, Y, zero_diagonal = _check_input(X, Y, zero_diagonal)
+
+    norm = torch.norm(X, p=2, dim=1)
+    X = X / norm.unsqueeze(1)
+    norm = torch.norm(Y, p=2, dim=1)
```

Comment:
norm of None?

Reply:
if `Y` is `None` it will be set to `X` during the `_check_input` function such that the computations are done between the rows of `X`